tiledb/sm/filesystem/s3.cc

```diff
@@ -203,7 +203,7 @@ S3::~S3() {
 /* ********************************* */
 
 Status S3::init(
-    stats::Stats* const parent_stats,
+    tdb_shared_ptr<stats::Stats> const parent_stats,
```

Comment:
Do we want the `tdb_shared_ptr` objects to be passed by const reference to avoid an object copy? If so, this change should propagate everywhere.

Reply:
Good question. So, if we pass a shared_ptr by reference we will indeed gain in performance but we will be kind of defeating the purpose of using a `shared_ptr`, as it will not increment the reference counting - thus will not share the ownership of the pointed object. If this is what we want though, we might as well use a raw pointer as before. Ideally what we should do is review all usages of `Stats` and decide where ownership should be shared and when not, and use accordingly pass-by-value (when needed) or pass-by-reference/raw pointer when it's not needed.

This is a great comment though because it made me think of [https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rr-smartptrparam](url) and actually doubt if this transition that I suggested to `shared_ptr` everywhere should happen for `Stats`.
I think the way it has been designed so far made sense: `Stats` are a tree owned as a `shared_ptr` by `Context` and `GlobalStats`, as their lifetime is defined by them: their lifetime actually it is the lifetime of the program as `GlobalStats` is a global variable, so I guess we want on purpose all stats to be available for dumping until the end. All other classes are just users of `Stats`, so passing around a raw pointer is just fine. 
@eric-hughes-tiledb what do you think?

Reply:
On the technical question, copying `shared_ptr` has quite low overhead on modern implementations. It was one of the early library functions to widely use a lock-free algorithm for performance. It's nothing to worry about.

For example, from [GCC docs](https://gcc.gnu.org/onlinedocs/libstdc++/manual/memory.html):

>The Boost shared_ptr (as used in GCC) features a clever lock-free algorithm to avoid the race condition, but this relies on the processor supporting an atomic Compare-And-Swap instruction.

Reply:
On an implicit engineering question, the principal point I want to make is that the technical implementation of whatever's done here should be readily maintainable. That means having a type symbol in the declaration, be that from a typedef, a using-declaration, or a class. If that's the case, then the implementation can evolve to meet technical goals without widespread changes each time.

From this perspective, this change set does two related things: (1) it creates a common convention for stats variables and (2) it standardizes around `shared_ptr` to accomplish it. Item (1) is more important to do than (2). If an implementation with `shared_ptr` does not have any serious problems, then I'd recommend just going forward with getting consistency in place and fix less-than-serious problems in a subsequent PR.

Reply:
And finally on the design question:

>their lifetime actually it is the lifetime of the program as GlobalStats is a global variable, so I guess we want on purpose all stats to be available for dumping until the end

If it's not documented that we want that lifetime, don't assume that it's the intent. It's far more likely that using a global was simply an early implementation convenience and shortcut that never got further attention. It seems like the wrong approach to use a global for all the standard reasons, in fact.

What we do want, well, we have to decide that. But we don't have to decide it in this PR.

Reply:
> but we will be kind of defeating the purpose of using a shared_ptr, as it will not increment the reference counting

I don't think this is true, because we don't care about incrementing the count upon copying the argument in the function, but when we copy the shared pointer to the class member. So, I think this is a non-issue here. Still the question remains: what convention / best practices should we use? And of course, whatever we decide, let's document in our new style guide :).

Reply:
^^ cc @eric-hughes-tiledb, unless I missed something here.